### PR TITLE
Update otelcontribcol to v0.106.0-gke.2

### PR DESCRIPTION
**Pull request by tiffanny29631**
Upgrading to latest for vulnerability fix.

To upgrade to otelcontribcol 0.106.0 we need to mitigate two breaking changes in https://github.com/open-telemetry/opentelemetry-collector-contrib/releases/tag/v0.104.0.

The opencensus receiver depends on 0.0.0.0 so disabling feature gate component.UseLocalHostAsDefaultHost.

With style `$FOO` deprecation, all env vars are changed to match the style. This brings a problem in the otel-agent container start when reconciler-manager and resource-group-controller do not initialize the following env vars:

- `configsync_sync_kind`
- `configsync_sync_name`
- `configsync_sync_namespace`

So splitting the otel-agent config between the reconciler and controllers.

### CODE REVIEW

File: Makefile
Comment (sdowell) [inline Makefile:1]: nit: typo in commit message (`veresion ` -> `version`)

Comment (tiffanny29631) [reply Makefile:1]: Fixed

**Review by sdowell — APPROVED** [review]
/lgtm

**Review by nan-yu — APPROVED** [review]
/lgtm

**Comment by tiffanny29631** [comment]
The failure is due to the two `very important breaking change` in the https://github.com/open-telemetry/opentelemetry-collector-contrib/releases/tag/v0.104.0. 

For the first one we could disable the feature gate as an equivalent of using 0.0.0.0.

The second one can be mitigated with a temporary feature gate disable, but still requires change in ConfigMap to match the style when the previous format is deprecated.

File: manifests/otel-agent-reconciler-cm.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Copyright 2022 Google LLC
```
Comment (sdowell) [inline manifests/otel-agent-reconciler-cm.yaml:1]: 2024

File: manifests/otel-agent-reconciler-cm.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    configmanagement.gke.io/arch: "csmr"
+data:
+  otel-agent-reconciler-config.yaml: |
+    receivers:
+      opencensus:
+    exporters:
+      opencensus:
+        endpoint: otel-collector.config-management-monitoring:55678
+        tls:
+          insecure: true
```
Comment (sdowell) [inline manifests/otel-agent-reconciler-cm.yaml:33]: What is the justification for disabling TLS?

Comment (tiffanny29631) [reply manifests/otel-agent-reconciler-cm.yaml:33]: This exporter in otel-agent sends data to the otel-collector receivers that runs on the same local host, so TLS is likely not necessary.

File: manifests/otel-agent-reconciler-cm.yaml (hunk 1 above)
Comment (tiffanny29631) [reply manifests/otel-agent-reconciler-cm.yaml]: Done

File: manifests/otel-agent-reconciler-cm.yaml (hunk 2 above)
Comment (sdowell) [reply manifests/otel-agent-reconciler-cm.yaml:33]: I'm not super familiar with these configurations - but it looks like the endpoint referenced in this block is for another Service running in the cluster, which may not be on the same local host. Am I missing something?

Comment (karlkfi) [reply manifests/otel-agent-reconciler-cm.yaml:33]: This looks like the sidecar agent config for the exporter sending to the otel-collector service/deployment.

That said, does the otel-collector have a TLS cert generated and injected or is it just using http?

Comment (karlkfi) [reply manifests/otel-agent-reconciler-cm.yaml:33]: If it's supposed to be the opencensus config for the reconciler writing to the sidecar agent, then the endpoint it wrong. It should be localhost. Otherwise, you're bypassing the agent.

Comment (tiffanny29631) [reply manifests/otel-agent-reconciler-cm.yaml:33]: Ah yeah sorry, it's sending to a different pod. In this case
- our metrics are custom metric data which are not critical to the healthiness of k8s or our product
- encryption may introduce extra overhead, and the cert configuration is an extra step in UX (or the certs need to be shipped as well)
- (this part I'm not sure) e2e tests are verifying metrics for every sync operation, the encryption may add work to the verification

Comment (karlkfi) [reply manifests/otel-agent-reconciler-cm.yaml:33]: FWIW, it looks like this section is just a copy of the existing agent config, which is already exporting metrics with TLS disabled.

IMO, we should probably enable TLS, but that would be a different feature/project.

So what we have here is fine for now.

Comment (tiffanny29631) [reply manifests/otel-agent-reconciler-cm.yaml:33]: A copy it is indeed :]

File: manifests/templates/resourcegroup-manifest.yaml [hunk 3]
```diff
@@ -231,6 +231,8 @@ spec:
             - ALL
       - args:
         - --config=/conf/otel-agent-config.yaml
+        - --feature-gates=-component.UseLocalHostAsDefaultHost
+        - --feature-gates=-pkg.translator.prometheus.NormalizeName
```
Comment (karlkfi) [inline manifests/templates/resourcegroup-manifest.yaml:235]: What's pkg.translator.prometheus.NormalizeName for?

File: manifests/otel-agent-cm.yaml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2022 Google LLC
+# Copyright 2024 Google LLC
```
Comment (karlkfi) [inline manifests/otel-agent-cm.yaml:1]: nit: google policy on copyright headers is to set it when the file is created and leave it. So it doesn't need to be updated here.

File: Makefile [hunk 5]
```diff
@@ -69,14 +69,14 @@ KUSTOMIZE_VERSION := v5.4.2-gke.0
 KUSTOMIZE := $(BIN_DIR)/kustomize
 KUSTOMIZE_STAGING_DIR := $(OUTPUT_DIR)/third_party/kustomize
 
-HELM_VERSION := v3.14.4-gke.2
+HELM_VERSION := v3.15.3-gke.0
```
Comment (karlkfi) [inline Makefile:72]: Can you move the helm update to a different PR?
Best to keep these seperate in case we need to revert.

File: manifests/templates/resourcegroup-manifest.yaml (hunk 3 above)
Comment (tiffanny29631) [reply manifests/templates/resourcegroup-manifest.yaml:235]: This is for dropping the feature gate AKA disabling it, to revert to a simple fashion of name sanitizing, I realized metrics from reconilers and reconciler-manager are using this but not the resource-group-controller so adding it here. 

See https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/translator/prometheus/README.md

File: Makefile (hunk 5 above)
Comment (tiffanny29631) [reply Makefile]: Done, modified commit and title.

File: manifests/otel-agent-cm.yaml (hunk 4 above)
Comment (tiffanny29631) [reply manifests/otel-agent-cm.yaml]: ack

File: manifests/templates/resourcegroup-manifest.yaml (hunk 3 above)
Comment (karlkfi) [reply manifests/templates/resourcegroup-manifest.yaml:235]: Can you make this specific change in a followup PR as well? Seems like a bug fix. Need to explain what it actually changes and what was broken.

Comment (tiffanny29631) [reply manifests/templates/resourcegroup-manifest.yaml]: Just removed the flag. Took a closer look - resource group controller only has gauge and distribution metrics, and the names are only composed with letters and underscores, so even with the normalizer on, no conversion is happening. Plus the test cases are not checking the r-g-c metrics, so no observations if there's a mismatch in name somehow.

We could add the flag just to be safe and align with other deployments.

Comment (karlkfi) [reply manifests/templates/resourcegroup-manifest.yaml]: Feel free to add it in a new PR just for consistency.

**Review by karlkfi — APPROVED** [review]
/lgtm

File: manifests/templates/resourcegroup-manifest.yaml (hunk 3 above)
Comment (tiffanny29631) [reply manifests/templates/resourcegroup-manifest.yaml]: Ohhhh, I see, change was made a while back to actually enable this feature, but adding a `-` disabled it instead. https://github.com/GoogleContainerTools/kpt-config-sync/pull/922/commits/f3e2a5e1b371855afa83f89f009e2befd988bcdd

To re-enable it we could just remove the feature-gate flag as the feature is default on. But the trim & append will start to work and e2e test would fail (just ran a quick test).

From my reading of the doc it'd be better to have the normalizer running to convert the names for the favor of the Prometheus exporter/importer, which would break our e2e by a bit as it explicitly looks for the metric names defined and registered upon creation.

I lean keeping it as is and ignoring the naming, since any user with Prometheus operator would probably already hooked with the current naming.

Or we have a thorough & proper enablement of the feature.

WDYT @karlkfi
